server/timeout_test.go

```diff
@@ -0,0 +1,67 @@
+package server
+
+import (
+	"crypto/rand"
+	"math/big"
+	"testing"
+	"time"
+)
+
+func TestTimeoutManager(t *testing.T) {
+	tm := newTimeoutManager()
+
+	// test 0 timeout means timeout does not occur
+	tm.SetTimeout(0)
+
+	// test fuzzing - 0 timeout - multiple sequential calls to random init and stop funcs
+	for i := 0; i < 30; i++ {
+		b, err := rand.Int(rand.Reader, big.NewInt(2))
+		if err != nil {
+			t.Error(err)
+		}
+
+		if b.Int64() == 1 {
+			tm.StartTimeout(t.FailNow)
+		} else {
+			tm.StopTimeout()
+		}
+	}
+
+	// test fuzzing - random timeout - multiple sequential calls to random init and stop funcs
+	for i := 0; i < 30; i++ {
+		b, err := rand.Int(rand.Reader, big.NewInt(2))
+		if err != nil {
+			t.Error(err)
+		}
+		to, err := rand.Int(rand.Reader, big.NewInt(11))
+		if err != nil {
+			t.Error(err)
+		}
+
+		tm.SetTimeout(uint(to.Int64() * 10))
+
+		if b.Int64() == 1 {
+			tm.StartTimeout(t.FailNow)
```

Comment:
this PR looks good except this test, especially this for loop, I was wondering if there is a possible that it will take up > `uint(to.Int64() * 10)` so that this test may fail 🙂

Reply:
The minimum timeout will be 10 milliseconds which is a long time for a function like this to run. However, you have highlighted an important point 💡.

If the timeout is 10 milliseconds and is the final call in the loop, it is possible that the `t.FailNow` could be called before line 52 is executed. So to prevent this I will add a `tm.StopTimeout()` between line 47 and 48.

The purpose of this loop is to flood the timeout manager with multiple random calls. If the process blocks or hangs anywhere `FailNow` should be triggered.

I could increase the timeout multiplier from `10` to `20` milliseconds, meaning that the loop has 20 - 200 milliseconds before a `FailNow`. That is a **lot** of time for this sequence to execute, if it fails under this circumstance something has blocked, deadlocked or the machine running the test is really really slow.

What do you think @qfrank ?

Reply:
sounds good 🫶

Reply:
I played around with this and it seems that at least on my machine 2 milliseconds is safe, 1 millisecond is not safe. I played around with the below code.

```go
if b.Int64() == 1 {
	tm.StartTimeout(t.FailNow)
	time.Sleep(time.Duration(tm.timeout-2) * time.Millisecond)
} else {
	tm.StopTimeout()
}
```

Another thing I noticed without the sleep calls, the entire test takes ~20 milliseconds to run, with the sleeps it takes ~2000 milliseconds. So that suggests to me that unless a machine is more than 100 times slower than my machine there won't be problems.

But I do take the point that this test in strict terms is flaky and none deterministic.